src/name.cpp

```diff
@@ -1,154 +1,191 @@
 #include "name.h"
 
 #include <cstddef>
-#include <map>
-#include <string>
 #include <utility>
-#include <vector>
 
 #include "cata_utility.h"
 #include "json.h"
 #include "rng.h"
 #include "string_formatter.h"
 #include "translations.h"
+#include "path_info.h"
 
-namespace Name
-{
-static std::map< nameFlags, std::vector< std::string > > names;
-
-static const std::map< std::string, nameFlags > usage_flags = {
-    { "given",     nameIsGivenName },
-    { "family",    nameIsFamilyName },
-    { "universal", nameIsGivenName | nameIsFamilyName },
-    { "nick",      nameIsNickName },
-    { "backer",    nameIsFullName },
-    { "city",      nameIsTownName },
-    { "world",     nameIsWorldName }
-};
-
-static const std::map< std::string, nameFlags > gender_flags {
-    { "male",   nameIsMaleName },
-    { "female", nameIsFemaleName },
-    { "unisex", nameIsUnisexName }
-};
-
-static nameFlags usage_flag( const std::string &usage )
+/*
+ * Nullptr, because instance will be initialized on demand.
+ */
+RandomName* RandomName::instance = nullptr;
+
+RandomName::RandomName( )
 {
-    const auto it = usage_flags.find( usage );
-    if( it != usage_flags.end() ) {
-        return it->second;
-    }
-    return static_cast< nameFlags >( 0 );
+    loadNames("");
 }
 
-static nameFlags gender_flag( const std::string &gender )
+RandomName::RandomName( std::string pathToFile )
 {
-    const auto it = gender_flags.find( gender );
-    if( it != gender_flags.end() ) {
-        return it->second;
-    }
-    return static_cast< nameFlags >( 0 );
+    loadNames( pathToFile );
 }
 
-// The loaded name is one of usage with optional gender.
-// The combinations used in names files are as follows.
-//
-// Backer | (Female|Male|Unisex)
-// Given  | (Female|Male)        // unisex names are duplicated in each group
-// Family | Unisex
-// Nick
-// City
-// World
-static void load( JsonIn &jsin )
+void RandomName::loadNames( std::string pathToFile )
 {
-    jsin.start_array();
+    std::string pathToFileJSON = pathToFile;
+
+    if (pathToFileJSON.empty())
+    {
+        Path *path = Path::getInstance();
```

Comment:
It seems your `getInstance` functions always return a valid pointer (you do not check for it being null and instead dereference them unconditionally). So if they do indeed never return an invalid pointer, why are they not returning a reference instead?

Reply:
## I'll make a brief definition of the reason for the change:

#### What was the behavior of _FILENAMES_?

Be a global access point (as it was a global variable).

#### Behavior of the class {Path}:

To be a global access point, I achieve it using the _Singleton_ pattern. 

------

#### Functionality of _FILENAMES_:

Provide the routes of the different pre-fabricated (assets) and configuration files of the App, this is done directly with FILENAMES.

#### Class {Path} functionality:

Provide the paths of the various pre-fabricated (assets) and configuration files of the App, I use a class to perform these tasks.

-------

### Which we currently have:

I have replaced the direct uses of FILENAMES with the call to the corresponding method of the Path instance, this means to use the _Singleton_ pattern, which makes sure of two things:

1. Imitate the behavior of global variable
2. Avoid creating more than one object

The very definition of the _Singleton_ pattern invites me to use pointers to imitate this feature of FILENAMES.

### Why don't you check the pointer validates

Theoretically should not fail the creation of the single instance (beyond extraordinary problems), a failure at this point means the impossibility of regaining control of the App. (mainly because the App. would not know how to find the paths to the files needed for execution).

### Why a reference is not returned instead

It's exactly what I haven't been able to do yet, initially, this was the first change I made to the App base code, and the initial approach was to use the _Singleton_ pattern to imitate the functionality that FILENAMES did, later I started refactoring another code and I forgot about it (to consider it finished), however there is a point I must say, this code section has a memory failure (**Memory Leak**) that I must correct and I have not done yet why its implementation brings other problems that I have not been able to solve.

The variant of the _Singleton_ pattern I use is the simplest and easiest to understand, but as I mentioned earlier has a memory failure (**Memory Leak**) and is Not Thread Safety, however there are modern versions of this pattern for versions of C++11 and C++14 that solve the memory failure, are Thread Safety and also return reference instead of pointers, I have to make this change for this pattern, here in this class and also in the other two {RandomName and OptionsManager}.

-------

**Disclaimers:**

The memory failure (**Memory Leak**) of the pattern has been checked by me, using Valgrind.

The Not Thread Safety of the pattern has been mentioned in the following article:
 
> C++ and the Perils of Double-Checked Locking
> Scott Meyers and Andrei Alexandrescu
> September 2004

------

_This has been made possible by Deepl._

Reply:
> Why a reference is not returned instead

Thread safety: a function static object (and even a global static object) are created in a thread safe manner:
```C++
Something &get_instance() {
	static Something the_instance;
	return the_instance;
}
```
And error detection would be done in the constructor, which in turn would just throw when anything bad happens. The next call to `get_instance` will attempt to create the instance again.

Or if you already have a pointer, return it as reference anyway:
```C++
Something *the_pointer = ...;

Something &get_instance() {
	if( !the_pointer ) {
		throw runtime_error(...);
	}
	return *the_pointer; // a reference now.
}
```